### PR TITLE
Improve accuracy for models using shuffle, unshuffle, cat ops (#19159)

### LINKED ISSUE #19176: Accuracy issue in QCOM backend for DSR models
Task to capture the accuracy debugging for qcom et backend.

### PR DESCRIPTION
**Pull request by abhinaykukkadapu**
Summary:

Replace the Qualcomm concat observer path with an explicit same-domain-or-requantize model for `aten.cat`. Preserve shared qparams for `pixel_shuffle` and `pixel_unshuffle`, extend `split_with_sizes_copy` coverage, and add regressions for mismatched `cat` branches plus value-preserving ops that must use `SharedQuantizationSpec`.

Differential Revision: D102626539

### CODE REVIEW

File: backends/qualcomm/quantizer/annotators/htp_rules.py [hunk 1]
```diff
@@ -234,35 +230,23 @@ def annotate(node: Node, quantization_config: QuantizationConfig) -> None:
...
+            assert isinstance(input_node, Node)
+            if input_node not in input_qspec_map:
+                input_qspec_map[input_node] = quantization_config.input_activation
 
-        output_qspec = QuantizationSpec(
-            dtype=quantization_config.output_activation.dtype,
-            qscheme=quantization_config.output_activation.qscheme,
-            quant_max=quantization_config.output_activation.quant_max,
-            quant_min=quantization_config.output_activation.quant_min,
-            observer_or_fake_quant_ctr=ConcatObserver.with_args(
```
Comment (winskuo-quic) [inline backends/qualcomm/quantizer/annotators/htp_rules.py:253]: I believe this change it reverting what this PR is doing: https://github.com/pytorch/executorch/pull/15162.
The reason https://github.com/pytorch/executorch/pull/15162 is introduced is because the input[0] could not cover the entire range of values for concat output, so a lot of output values were clipped.

If you have 2 input tensors like:
`sample_input = (
            torch.tensor([[[[-10.0, 2.0], [3.0, 4.0]]]]),
            torch.tensor([[[[1.0, 3.0], [8.0, 10.0]]]]),
        )`
and after it goes through cat operation, you will be getting the wrong value with this PR.
`[tensor([[[[-9.9798,  1.9849],
          [ 2.9774,  4.0250],
          [ 1.0476,  3.0325],
          [ 4.0802,  4.0802]]]])]
`
I have a demo PR to reproduce this error, please have a look:
https://github.com/pytorch/executorch/pull/19182

File: backends/qualcomm/_passes/annotate_quant_attrs.py [hunk 2]
```diff
@@ -79,11 +85,56 @@ def _find_last_dq_nodes(self, node: torch.fx.node.Node) -> torch.fx.node.Node:
...
+            requant_attrs[QCOM_ENCODING] = source_q_attrs[QCOM_ENCODING]
+            source_node.meta.setdefault(QCOM_REQUANTIZE, {})
+            source_node.meta[QCOM_REQUANTIZE][cat_node.name] = requant_attrs
+
     def _annotate_requant(self, n):
         # Record requant attributes:
         # node1 -> q_ui8 (n) -> dq_ui8 -> q_int32 -> dq_int32 -> node2 -> ....
         # We store {node2: quant_attr in dq_int32} in node1.meta
         if n.target in q_ops and n.args[0].target not in dq_ops:
+            self._annotate_cat_requant(n)
```
Comment (winskuo-quic) [inline backends/qualcomm/_passes/annotate_quant_attrs.py]: I think adding a specific op in a general pass would not be the best option. Do you think we can possibly move the logic to `htp_rules.py` if this is necessary?

**Review by winskuo-quic — COMMENTED** [review]
Hi @abhinaykukkadapu,
Thanks a lot for this PR and making all changes to improve QNN Backend accuracy.
I think shuffle & unshuffle LGTM.
However, I think there might be some concerns for the concat operation modification.
I have a PR to reproduce the concerns I got: https://github.com/pytorch/executorch/pull/19182. I also have the outputs and graphs compared in the summary section.
Please have a look.

If change for concat behavior is required to fix accuracy issues for the model you mentioned in https://github.com/pytorch/executorch/pull/19159, could we possibly reuse `custom_annotation` feature under https://github.com/pytorch/executorch/blob/main/backends/qualcomm/quantizer/custom_annotation.py for that model.

Thanks

File: backends/qualcomm/quantizer/annotators/htp_rules.py (hunk 1 above)
Comment (abhinaykukkadapu) [reply backends/qualcomm/quantizer/annotators/htp_rules.py:253]: Thanks @winskuo-quic for the detailed review, i agree that it might've worked for the model but might not work when the ranges skewed like in your example. Let me revert the cat to concatobserver and test the accuracy.

File: backends/qualcomm/_passes/annotate_quant_attrs.py (hunk 2 above)
Comment (abhinaykukkadapu) [reply backends/qualcomm/_passes/annotate_quant_attrs.py]: Yeah, i was also going back and forth on this, i added a pass instead of shoving it up here.

**Comment by abhinaykukkadapu** [comment]
@winskuo-quic here is the latest patch, it uses the same ConcatObserver approach as previously done. Additionally the pass for requant is introduced after the generic AnnotateQuantAttr pass. Please review when you get a chance.

FYI the SQNR right now stayed above target with this changeset. (> 30db)

**Comment by abhinaykukkadapu** [comment]
@winskuo-quic @shewu-quic bumping this up in your queue.

**Comment by winskuo-quic** [comment]
> @winskuo-quic @shewu-quic bumping this up in your queue.

@abhinaykukkadapu Could you share more about how to see the difference with the repro script? https://github.com/pytorch/executorch/pull/19179
An example script will be appreciated.

I tried running the script with this PR but I am getting the same numbers both before and after disabling `AnnotateConcatRequant`
image

image

**Comment by abhinaykukkadapu** [comment]
> > @winskuo-quic @shewu-quic bumping this up in your queue.
> 
> @abhinaykukkadapu Could you share more about how to see the difference with the repro script? #19179 An example script will be appreciated.
> 
> I tried running the script with this PR but I am getting the same numbers both before and after disabling `AnnotateConcatRequant` image
> 
> image

This PR is unrelated to the repro script, as i mentioned in the email, that is for an issue i couldn't solve outside QNN, it is to reproduce an issue that happened to those specific sequence of ops, see the `quant_vs_runtime_sqnr`

What this PR actually does is to fix the wrong annotation for shuffle, unshuffle and cat ops requant (we need requant even though we use observer which is missing). Lmk if you need more info.

**Comment by winskuo-quic** [comment]
> This PR is unrelated to the repro script, as i mentioned in the email, that is for an issue i couldn't solve outside QNN, it is to reproduce an issue that happened to those specific sequence of ops, see the `quant_vs_runtime_sqnr`
> 
> What this PR actually does is to fix the wrong annotation for shuffle, unshuffle and cat ops requant (we need requant even though we use observer which is missing). Lmk if you need more info.

Sorry I think I misunderstood your comments above where you mentioned SQNR got an increase with the Concat fix, so I thought updating quant configs for Concat can improve: https://github.com/pytorch/executorch/pull/19179

I think I am still unable to see the difference with and without `AnnotateConcatRequant`. I tried running `test_qnn_backend_cat_uses_concat_observer_output_qspec` and still can't see difference in final graph.
Could you please provide a simple test or the model you are enabling that can show the difference before and after enabling `AnnotateConcatRequant`, e.g. `without pass -> accuracy fail` and `with pass -> accuracy pass`.  This would also serve as a UT to show that `AnnotateConcatRequant` is necessary.
Thanks

**Comment by abhinaykukkadapu** [comment]
> > This PR is unrelated to the repro script, as i mentioned in the email, that is for an issue i couldn't solve outside QNN, it is to reproduce an issue that happened to those specific sequence of ops, see the `quant_vs_runtime_sqnr`
> > What this PR actually does is to fix the wrong annotation for shuffle, unshuffle and cat ops requant (we need requant even though we use observer which is missing). Lmk if you need more info.
> 
> Sorry I think I misunderstood your comments above where you mentioned SQNR got an increase with the Concat fix, so I thought updating quant configs for Concat can improve: #19179
> 
> I think I am still unable to see the difference with and without `AnnotateConcatRequant`. I tried running `test_qnn_backend_cat_uses_concat_observer_output_qspec` and still can't see difference in final graph. Could you please provide a simple test or the model you are enabling that can show the difference before and after enabling `AnnotateConcatRequant`, e.g. `without pass -> accuracy fail` and `with pass -> accuracy pass`. This would also serve as a UT to show that `AnnotateConcatRequant` is necessary. Thanks

Thanks for the clarification and sorry for the confusion. The SQNR comment i made was the test i made with the real DSR model. TLDR is that the requant is needed for cat ops when there is a shared spec in one of the input branches, in that case annotation itself is not sufficient and we need to requant after convert_pt2e and is done by the pass.

Added unit tests to simulate the low and high accuracy with and without the pass.

**Comment by winskuo-quic** [comment]
Hi @abhinaykukkadapu,
Thanks for all the changes. I have checked the PR UT but still observer both cases (with and without annotate_concat_requant passes) generates incorrect output compared to nn.Module FP32.
Reference:
image

I think shuffle and unshuffle looks good to me. Do you think we should merge these 2 first and maybe we can take a look at concat quantization to see if there is a more general approach to cover all values?
Thanks

**Comment by abhinaykukkadapu** [comment]
> Do you think we should merge these 2 first and maybe we can take a look at concat quantization to see if there is a more general approach to cover all values?

@winskuo-quic sounds good, updated the PR.

**Review by winskuo-quic — APPROVED** [review]
LGTM. Thanks for the effot!

**Review by psiddh — APPROVED** [review]
